staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -0,0 +1,170 @@
+	fq "k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing"
+	fcmetrics "k8s.io/apiserver/pkg/util/flowcontrol/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+const (
+	decisionExecute = iota
+	decisionCancel
+)
+
+type fakeApfFilter struct {
+	mockDecision int
+}
+
+func (t fakeApfFilter) Handle(ctx context.Context,
+	requestDigest utilflowcontrol.RequestDigest,
+	noteFn func(fs *fctypesv1a1.FlowSchema, pl *fctypesv1a1.PriorityLevelConfiguration),
+	queueNoteFn fq.QueueNoteFn,
+	execFn func(),
+) {
+	noteFn(bootstrap.SuggestedFlowSchemaGlobalDefault, bootstrap.SuggestedPriorityLevelConfigurationGlobalDefault)
+	switch t.mockDecision {
+	case decisionExecute:
+		queueNoteFn(true)
+		execFn()
+	case decisionCancel:
+		queueNoteFn(true)
+	}
+}
+
+func (t fakeApfFilter) Run(stopCh <-chan struct{}) error {
+	return nil
+}
+
+func (t fakeApfFilter) Install(c *mux.PathRecorderMux) {
+}
+
+func newApfDecisionHandler(decision int) http.Handler {
+	longRunningRequestCheck := BasicLongRunningRequestCheck(sets.NewString("watch"), sets.NewString("proxy"))
+	apfHandler := WithPriorityAndFairness(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+	}), longRunningRequestCheck, fakeApfFilter{mockDecision: decision})
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		r = r.WithContext(apirequest.WithUser(r.Context(), &user.DefaultInfo{
+			Groups: []string{user.AllUnauthenticated},
+		}))
+		apfHandler.ServeHTTP(w, r)
+	})
+}
+
+func newApfServer(apfHandler http.Handler) (*httptest.Server, error) {
+	requestInfoFactory := &apirequest.RequestInfoFactory{APIPrefixes: sets.NewString("apis", "api"), GrouplessAPIPrefixes: sets.NewString("api")}
+	handler := apifilters.WithRequestInfo(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		r = r.WithContext(apirequest.WithUser(r.Context(), &user.DefaultInfo{
+			Groups: []string{user.AllUnauthenticated},
+		}))
+		apfHandler.ServeHTTP(w, r)
+	}), requestInfoFactory)
+
+	apfServer := httptest.NewServer(handler)
+	return apfServer, nil
+}
+
+func TestApfSkipLongRunningRequest(t *testing.T) {
+	epmetrics.Register()
+	apfHandler := newApfDecisionHandler(decisionCancel)
+
+	server, err := newApfServer(apfHandler)
+	assert.NoError(t, err)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces?watch=true", server.URL), http.StatusOK); err != nil {
+		// request should not be rejected
+		t.Error(err)
+	}
+	assert.NoError(t, err)
+}
+
+func TestApfCanceledRequest(t *testing.T) {
+	epmetrics.Register()
+	fcmetrics.Register()
+	apfHandler := newApfDecisionHandler(decisionCancel)
+
+	server, err := newApfServer(apfHandler)
+	assert.NoError(t, err)
+	defer server.Close()
+
+	go func() {
+		if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusTooManyRequests); err != nil {
+			t.Error(err)
+		}
+	}()
+	assert.NoError(t, err)
+
+	//wait metrics to be collected
+	time.Sleep(time.Second)
```

Comment:
Waiting exactly one second produces a race.  I would wait longer.

Reply:
Set the duration to 1.5s.